### PR TITLE
[SPARK-42454][SQL] SPJ: encapsulate all SPJ related parameters in BatchScanExec

### PR DESCRIPTION
**Pull request by szehon-ho**
### What changes were proposed in this pull request?
Pull out the SPJ-related attribute of BatchScanExec into a case class 

### Why are the changes needed?
We plan to have further evolution of SPJ parameters to support more SPJ features.  So we want to stabilize the definition of BatchScanExec to not have to touch the many places in the code that it is pattern-matched/unapplied, etc..

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Existing unit test to verify no behavior change.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/BatchScanExec.scala [hunk 1]
```diff
@@ -37,23 +37,19 @@ case class BatchScanExec(
...
     ordering: Option[Seq[SortOrder]] = None,
     @transient table: Table,
-    commonPartitionValues: Option[Seq[(InternalRow, Int)]] = None,
-    applyPartialClustering: Boolean = false,
-    replicatePartitions: Boolean = false) extends DataSourceV2ScanExecBase {
+    spjParams: StoragePartitionJoinParams = StoragePartitionJoinParams()
+  ) extends DataSourceV2ScanExecBase {
 
-  @transient lazy val batch = if (scan == null) null else scan.toBatch
+  @transient lazy val batch: Batch = if (scan == null) null else scan.toBatch
```
Comment (szehon-ho) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/BatchScanExec.scala:45]: This is just to fix the warning, not related , lmk if we want to do it in another pr

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/BatchScanExec.scala [hunk 2]
```diff
@@ -148,15 +144,17 @@ case class BatchScanExec(
                   s"${SQLConf.V2_BUCKETING_PARTIALLY_CLUSTERED_DISTRIBUTION_ENABLED.key} " +
                   "is enabled")
 
-            val groupedPartitions = groupPartitions(finalPartitions.map(_.head), true).get
+            val groupedPartitions = groupPartitions(finalPartitions.map(_.head),
```
Comment (szehon-ho) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/BatchScanExec.scala:147]: same was for unrelated warning, and was thinking to fix as im in this code anyway

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/BatchScanExec.scala [hunk 3]
```diff
@@ -241,3 +243,24 @@ case class BatchScanExec(
...
+    case other: StoragePartitionJoinParams =>
+      this.commonPartitionValues == other.commonPartitionValues &&
+      this.replicatePartitions == other.replicatePartitions &&
+      this.applyPartialClustering == other.applyPartialClustering
+    case _ =>
+      false
+  }
+
+  override def hashCode(): Int = Objects.hashCode(
+    commonPartitionValues: Option[Seq[(InternalRow, Int)]],
```
Comment (sunchao) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/BatchScanExec.scala:262]: nit: why do we need to explicitly specify the types here?

**Review by sunchao — APPROVED** [review]
LGTM, cc @cloud-fan too

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/BatchScanExec.scala (hunk 3 above)
Comment (szehon-ho) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/BatchScanExec.scala:262]: whoops, will remove

Comment (sunchao) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/BatchScanExec.scala:262]: I see, it is OK then.

**Comment by sunchao** [comment]
Thanks, merged to master!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/BatchScanExec.scala [hunk 4]
```diff
@@ -241,3 +243,24 @@ case class BatchScanExec(
     s"BatchScan ${table.name()}".trim
   }
 }
+
+case class StoragePartitionJoinParams(
+    keyGroupedPartitioning: Option[Seq[Expression]] = None,
+    commonPartitionValues: Option[Seq[(InternalRow, Int)]] = None,
+    applyPartialClustering: Boolean = false,
+    replicatePartitions: Boolean = false) {
+  override def equals(other: Any): Boolean = other match {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/BatchScanExec.scala:252]: This is not new in this PR, but it's great if we can add some comments about why we exclude `keyGroupedPartitioning` in the comparison. cc @sunchao

Comment (sunchao) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/BatchScanExec.scala:252]: sure will do @cloud-fan (sorry just saw this!).
